File: Common/Orders/Slippage/VolumeShareSlippageModel.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using QuantConnect.Data.Market;
using QuantConnect.Logging;
using QuantConnect.Securities;
using System;
using static QuantConnect.StringExtensions;

namespace QuantConnect.Orders.Slippage
{
    /// <summary>
    /// Represents a slippage model that is calculated by multiplying the price impact constant
    /// by the square of the ratio of the order to the total volume.
    /// </summary>
    public class VolumeShareSlippageModel : ISlippageModel
    {
        private readonly decimal _priceImpact;
        private readonly decimal _volumeLimit;

        /// <summary>
        /// Initializes a new instance of the <see cref="VolumeShareSlippageModel"/> class
        /// </summary>
        /// <param name="volumeLimit"></param>
        /// <param name="priceImpact">Defines how large of an impact the order will have on the price calculation</param>
        public VolumeShareSlippageModel(decimal volumeLimit = 0.025m, decimal priceImpact = 0.1m)
        {
            _priceImpact = priceImpact;
            _volumeLimit = volumeLimit;
        }

        /// <summary>
        /// Slippage Model. Return a decimal cash slippage approximation on the order.
        /// </summary>
        public decimal GetSlippageApproximation(Security asset, Order order)
        {
            var lastData = asset.GetLastData();
            if (lastData == null) return 0;

            var barVolume = 0m;
            var slippagePercent = _volumeLimit * _volumeLimit * _priceImpact;

            switch (lastData.DataType)
            {
                case MarketDataType.TradeBar:
                    barVolume = ((TradeBar)lastData).Volume;
                    break;
                case MarketDataType.QuoteBar:
                    barVolume = order.Direction == OrderDirection.Buy
                        ? ((QuoteBar)lastData).LastBidSize
                        : ((QuoteBar)lastData).LastAskSize;
                    break;
                default:
                    throw new InvalidOperationException(
                        $"VolumeShareSlippageModel.GetSlippageApproximation: Cannot use this model with market data type {lastData.GetType()}"
                    );
            }

            // If volume is zero or negative, we use the maximum slippage percentage since the impact of any quantity is infinite
            // In FX/CFD case, we issue a warning and return zero slippage
            if (barVolume <= 0)
            {
                var securityType = asset.Symbol.ID.SecurityType;
                if (securityType == SecurityType.Cfd || securityType == SecurityType.Forex || securityType == SecurityType.Crypto)
                {
                    Log.Error(Invariant($"VolumeShareSlippageModel.GetSlippageApproximation: {securityType} security type often ") +
                        "does not report volume. If you intend to model slippage beyond the spread, please consider another model."
                    );
                    return 0;
                }

                Log.Error("VolumeShareSlippageModel.GetSlippageApproximation: Bar volume cannot be zero or negative. " +
                    Invariant($"Volume: {barVolume}. Using maximum slippage percentage of {slippagePercent}")
                );
            }
            else
            {
                // Ratio of the order to the total volume
                var volumeShare = Math.Min(order.AbsoluteQuantity / barVolume, _volumeLimit);

                slippagePercent = volumeShare * volumeShare * _priceImpact;
            }

            return slippagePercent * lastData.Value;
        }
    }
}

File: Common/Orders/Slippage/VolumeShareSlippageModel.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from AlgorithmImports import *

class VolumeShareSlippageModel:
    '''Represents a slippage model that is calculated by multiplying the price impact constant by the square of the ratio of the order to the total volume.'''
    
    def __init__(self, volume_limit: float = 0.025, price_impact: float = 0.1) -> None:
        '''Initializes a new instance of the "VolumeShareSlippageModel" class
        Args:
            volume_limit:
            price_impact: Defines how large of an impact the order will have on the price calculation'''
        self.volume_limit = volume_limit
        self.price_impact = price_impact

    def get_slippage_approximation(self, asset: Security, order: Order) -> float:
        '''Slippage Model. Return a decimal cash slippage approximation on the order.
        Args:
            asset: The Security instance of the security of the order.
            order: The Order instance being filled.'''
        last_data = asset.get_last_data()
        if not last_data:
           return 0

        bar_volume = 0
        slippage_percent = self.volume_limit * self.volume_limit * self.price_impact

        if last_data.data_type == MarketDataType.TRADE_BAR:
            bar_volume = last_data.volume
        elif last_data.data_type == MarketDataType.QUOTE_BAR:
            bar_volume = last_data.last_bid_size if order.direction == OrderDirection.BUY else last_data.last_ask_size
        else:
           raise InvalidOperationException(Messages.VolumeShareSlippageModel.InvalidMarketDataType(last_data))

        # If volume is zero or negative, we use the maximum slippage percentage since the impact of any quantity is infinite
        # In FX/CFD case, we issue a warning and return zero slippage
        if bar_volume <= 0:
            security_type = asset.symbol.id.security_type
            if security_type == SecurityType.CFD or security_type == SecurityType.FOREX or security_type == SecurityType.CRYPTO:
                Log.error(Messages.VolumeShareSlippageModel.VolumeNotReportedForMarketDataType(security_type))
                return 0

            Log.error(Messages.VolumeShareSlippageModel.NegativeOrZeroBarVolume(bar_volume, slippage_percent))
        else:
            # Ratio of the order to the total volume
            volume_share = min(order.absolute_quantity / bar_volume, self.volume_limit)

            slippage_percent = volume_share * volume_share * self.price_impact

        return slippage_percent * last_data.Value;
